packages/flutter/lib/src/widgets/pop_scope.dart

```diff
@@ -0,0 +1,60 @@
+import 'framework.dart';
+import 'navigator.dart';
+import 'routes.dart';
+
+class PopScope extends StatefulWidget {
+  const PopScope({
+    Key? key,
+    required this.child,
+    required this.onPop,
+  })  : assert(child != null),
+        super(key: key);
+
+  final Widget child;
+
+  final VoidCallback? onPop;
+
+  @override
+  State<PopScope> createState() => _PopScopeState();
+}
+
+class _PopScopeState extends State<PopScope> {
+  ModalRoute<dynamic>? _route;
+
+  @override
+  void didChangeDependencies() {
+    super.didChangeDependencies();
+    if (widget.onPop != null) {
+      _route?.removeScopedPopCallback(widget.onPop!);
+    }
+    _route = ModalRoute.of(context);
+    if (widget.onPop != null) {
+      _route?.addScopedPopCallback(widget.onPop!);
+    }
+  }
+
+  @override
+  void didUpdateWidget(PopScope oldWidget) {
+    super.didUpdateWidget(oldWidget);
+    assert(_route == ModalRoute.of(context));
```

Comment:
unnecessary assert, may be some crash, it is fixed in https://github.com/flutter/flutter/pull/89226